### DISCUSSION TITLE
How to connect signals from another node?

### DISCUSSION THREAD

**Discussion by GuiSaiUwU** [niklas2902/py4godot#72] [Q&A] [answered]
Currently, the pop up is being called so I can select a file, but the method itself is not called, I tried to use a default python method and it asked me to pass an `py4godot.classes.core.Callable` instead, so I made that Test class, which doesnt get called, I tried to do `def __call__` but still the same error :(

```py
@gdclass
class owo(Node2D):
	def _ready(self) -> None:
		file_dialog: FileDialog = FileDialog.cast(self.get_node(NodePath.new2("FileDialog")))
		file_dialog.connect(StringName.new2("_on_file_dialog_file_selected"), Test())

		print("Calling the select file thing OwO")
		file_dialog.set_visible(True)
		print("Done UwU")
		file_dialog.set_visible(False)

	def _process(self, delta:float) -> None:
		pass


class Test(Callable):
	def _on_file_dialog_file_selected(self, path: String):  # Not being called
		print(f"Path: {path}")
```

![image](https://github.com/user-attachments/assets/365bba11-aff8-45f0-8ff7-21823804da53)

**Comment by niklas2902** [comment, marked as answer]
Hi,
I'm sorry to hear that.
I quickly went over it. It seems, for callable to be valid, you kind of have to use the methods new1 or new2 for construction. So you have to point to a method of a gdclass:
```python
file_dialog.connect(StringName.new2("file_selected"), Callable.new2(self, StringName.new2("_on_file_dialog_file_selected")))
```
I don't know your specific use case. But there is a bit of a simpler way for add. I added the signals as properties of the classes and you can connect them directly (keep in mind, the method you want to connect to has to be a gdclass):

```python
file_dialog.file_selected.connect(self._on_file_dialog_file_selected)
```

here is my full example code:

```python

from py4godot.methods import private
from py4godot.signals import signal, SignalArg
from py4godot.classes import gdclass
from py4godot.classes.core import Vector3, Callable, String, StringName, NodePath
from py4godot.classes.Node2D.Node2D import Node2D
from py4godot.classes.FileDialog.FileDialog import FileDialog

@gdclass
class owo(Node2D):
	def _ready(self) -> None:
		file_dialog: FileDialog = FileDialog.cast(self.get_node(NodePath.new2("FileDialog")))
		# Method 1
		#file_dialog.connect(StringName.new2("file_selected"), Callable.new2(self, StringName.new2("_on_file_dialog_file_selected")))

		# Method 2
		# Please keep in mind, that the method must be a method of a gdclass
		file_dialog.file_selected.connect(self._on_file_dialog_file_selected)

		print("Calling the select file thing OwO")
		file_dialog.set_visible(True)
		print("Done UwU")
		file_dialog.set_visible(False)

	def _process(self, delta:float) -> None:
		pass
	
	def _on_file_dialog_file_selected(self, path: String):  # Not being called
		print(f"Path: {path}")

```

By the way. I couldn't find `_on_file_dialog_file_selected`. So I switched it to `file_selected`. I hope, this was the signal you were looking for.
I hope this solves your problem. If you run into problems, feel free to ask

**Comment by GuiSaiUwU** [comment]
Thank you for the answer, also could you reply why I get an error when I try to make the variable file_dialog declared in the class scope?
```py
@gdclass
class owo(Node2D):
	file_dialog: FileDialog

	def _ready(self) -> None:
		pass
```

Error:
```
 test:1 - exec_class: Exception happened:  (User)
  test:1 - class to load:res://owo.py  (User)
  test:1 - b'res://owo.py'  (User)
  test:1 - exec_class: Exception happened: Traceback (most recent call last):
    File "py4godot/pluginscript_api/utils/annotations.pyx", line 151, in py4godot.pluginscript_api.utils.annotations.exec_class
    File "py4godot/pluginscript_api/utils/annotations.pyx", line 109, in py4godot.pluginscript_api.utils.annotations.collect_properties
    File "py4godot/pluginscript_api/utils/annotations.pyx", line 92, in py4godot.pluginscript_api.utils.annotations.generate_default_val
  AttributeError: type object 'py4godot.classes.FileDialog.FileDialog.FileDialog' has no attribute 'new0'
    (User)
 ```
 
 It looks like its an issue of how im declaring it 🤔

**Comment by niklas2902** [comment]
I'm sorry, I made the api a bit confusing. 
For all classes inside the `py4godot.classes.core` module, you have the method `new0,new1` , etc. for instantiating them. For all the other classes, that can be instantiated like FileDialog, you have the method `construct`.
So hopefully creating the FileDialog like this:
```python
my_file_dialog = FileDialog.constructor()
```  

Should do the trick. I hope this helps
